File: clients/macos/Sources/FuseOSCore/SignalClient.swift
import Foundation

/// What we know about a paired peer right now.
///
/// `publicKey` and `lanAddress` are what make a direct connection possible: the address
/// says where to dial, the key says who must answer. Both arrive from the control plane,
/// which is the only thing that can vouch for them.
public struct PeerPresence {
    public let online: Bool
    public let battery: Int?
    public let publicKey: String?
    public let lanAddress: String?

    public init(online: Bool, battery: Int?, publicKey: String?, lanAddress: String?) {
        self.online = online
        self.battery = battery
        self.publicKey = publicKey
        self.lanAddress = lanAddress
    }
}

/// A decoded `/signal` server frame (only the fields the dashboard needs).
private struct SignalEvent: Decodable {
    let type: String
    let deviceId: String?
    let battery: Int?
    let online: Bool?
    let publicKey: String?
    let lanAddress: String?
    let peers: [PeerCard]?
}

private struct PeerCard: Decodable {
    let deviceId: String
    let battery: Int?
    let online: Bool?
    let publicKey: String?
    let lanAddress: String?
}

/// Maintains the `/signal` WebSocket: authenticates with a `hello`, sends battery
/// heartbeats, and surfaces peer presence via callbacks. Fail-soft — a dropped
/// socket reconnects with a short delay and never crashes the app.
@MainActor
public final class SignalClient {
    /// Everything currently known about this device's paired peers.
    public private(set) var presence: [String: PeerPresence] = [:]

    public var onPresenceChanged: (([String: PeerPresence]) -> Void)?
    public var onPaired: (() -> Void)?

    /// Supplies this device's `ip:port` for peers to dial. The LAN listener binds an
    /// ephemeral port, so this is nil until it is up and changes across restarts — hence
    /// a closure rather than a stored value.
    public var lanAddressProvider: (() -> String?)?

    private var task: URLSessionWebSocketTask?
    private var token: String?
    private var deviceId: String?
    private var active = false
    private var heartbeat: Task<Void, Never>?

    public init() {}

    public func start(token: String, deviceId: String) {
        self.token = token
        self.deviceId = deviceId
        active = true
        connect()
    }

    public func stop() {
        active = false
        heartbeat?.cancel()
        heartbeat = nil
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
    }

    private func connect() {
        guard active, let token, let deviceId else { return }
        let socket = URLSession.shared.webSocketTask(with: Config.signalURL)
        task = socket
        socket.resume()

        var hello: [String: Any] = ["type": "hello", "token": token, "deviceId": deviceId]
        if let battery = Battery.currentPercent() { hello["battery"] = battery }
        if let lanAddress = lanAddressProvider?() { hello["lanAddress"] = lanAddress }
        send(hello)

        receiveLoop()
        startHeartbeat()
    }

    private func startHeartbeat() {
        heartbeat?.cancel()
        heartbeat = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 20_000_000_000)
                guard let self, self.active else { return }
                var frame: [String: Any] = ["type": "heartbeat"]
                if let battery = Battery.currentPercent() { frame["battery"] = battery }
                if let lanAddress = self.lanAddressProvider?() { frame["lanAddress"] = lanAddress }
                self.send(frame)
            }
        }
    }

    private func send(_ payload: [String: Any]) {
        guard
            let task,
            let data = try? JSONSerialization.data(withJSONObject: payload),
            let text = String(data: data, encoding: .utf8)
        else { return }
        task.send(.string(text)) { _ in }  // fail soft
    }

    private func receiveLoop() {
        Task { @MainActor in
            guard let socket = self.task else { return }
            do {
                let message = try await socket.receive()
                if case let .string(text) = message { self.handle(text) }
                if self.active { self.receiveLoop() }
            } catch {
                if self.active { self.scheduleReconnect() }
            }
        }
    }

    private func scheduleReconnect() {
        guard active else { return }
        task = nil
        Task { @MainActor in
            try? await Task.sleep(nanoseconds: 2_000_000_000)
            guard self.active else { return }
            self.connect()
        }
    }

    /// Internal rather than private so tests can feed real server frames straight in,
    /// without standing up a WebSocket.
    func handle(_ text: String) {
        guard
            let data = text.data(using: .utf8),
            let event = try? JSONDecoder().decode(SignalEvent.self, from: data)
        else { return }

        switch event.type {
        case "hello-ok":
            for peer in event.peers ?? [] {
                setPresence(
                    peer.deviceId, online: peer.online ?? true, battery: peer.battery,
                    publicKey: peer.publicKey, lanAddress: peer.lanAddress,
                )
            }
        case "peer-online", "peer-update":
            if let id = event.deviceId {
                setPresence(
                    id, online: true, battery: event.battery,
                    publicKey: event.publicKey, lanAddress: event.lanAddress,
                )
            }
        case "peer-offline":
            // Keep the key and address on the way down: the peer is unreachable now, but
            // the details are still valid when it comes back and save a round trip.
            if let id = event.deviceId {
                setPresence(id, online: false, battery: nil, publicKey: nil, lanAddress: nil)
            }
        case "paired":
            onPaired?()
        default:
            break
        }
    }

    private func setPresence(
        _ deviceId: String,
        online: Bool,
        battery: Int?,
        publicKey: String?,
        lanAddress: String?,
    ) {
        let existing = presence[deviceId]
        presence[deviceId] = PeerPresence(
            online: online,
            battery: battery ?? existing?.battery,
            publicKey: publicKey ?? existing?.publicKey,
            lanAddress: lanAddress ?? existing?.lanAddress,
        )
        onPresenceChanged?(presence)
    }
}

File: clients/android/app/src/main/java/com/fuseos/app/data/SignalClient.kt
package com.fuseos.app.data

import io.ktor.client.HttpClient
import io.ktor.client.plugins.websocket.webSocket
import io.ktor.websocket.Frame
import io.ktor.websocket.readText
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * What we know about a paired peer right now.
 *
 * [publicKey] and [lanAddress] are what make a direct connection possible: the address
 * says where to dial, the key says who must answer. Both arrive from the control plane,
 * which is the only thing that can vouch for them.
 */
data class PeerPresence(
    val online: Boolean,
    val battery: Int?,
    val publicKey: String?,
    val lanAddress: String?,
)

@Serializable
private data class Hello(
    val type: String = "hello",
    val token: String,
    val deviceId: String,
    val battery: Int? = null,
    val lanAddress: String? = null,
)

@Serializable
private data class Heartbeat(
    val type: String = "heartbeat",
    val battery: Int? = null,
    val lanAddress: String? = null,
)

@Serializable
private data class SignalEvent(
    val type: String,
    val deviceId: String? = null,
    val battery: Int? = null,
    val online: Boolean? = null,
    val publicKey: String? = null,
    val lanAddress: String? = null,
    val peers: List<PeerCard>? = null,
)

@Serializable
private data class PeerCard(
    val deviceId: String,
    val battery: Int? = null,
    val online: Boolean? = null,
    val publicKey: String? = null,
    val lanAddress: String? = null,
)

/**
 * Maintains the `/signal` WebSocket: authenticates with a `hello`, sends battery
 * heartbeats, and exposes peer presence as a flow. Fail-soft — a dropped socket
 * reconnects after a short delay and never crashes the app.
 */
class SignalClient(
    private val client: HttpClient,
    private val signalUrl: String,
    private val scope: CoroutineScope,
    private val batteryProvider: () -> Int?,
    private val lanAddressProvider: () -> String? = { null },
) {
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }

    private val _presence = MutableStateFlow<Map<String, PeerPresence>>(emptyMap())
    val presence: StateFlow<Map<String, PeerPresence>> = _presence.asStateFlow()

    private val _paired = MutableSharedFlow<Unit>(extraBufferCapacity = 4)
    val paired: SharedFlow<Unit> = _paired.asSharedFlow()

    private var job: Job? = null

    fun start(token: String, deviceId: String) {
        stop()
        job = scope.launch { loop(token, deviceId) }
    }

    fun stop() {
        job?.cancel()
        job = null
        _presence.value = emptyMap()
    }

    private suspend fun loop(token: String, deviceId: String) {
        while (currentCoroutineContext().isActive) {
            try {
                client.webSocket(signalUrl) {
                    send(
                        Frame.Text(
                            json.encodeToString(
                                Hello(
                                    token = token,
                                    deviceId = deviceId,
                                    battery = batteryProvider(),
                                    lanAddress = lanAddressProvider(),
                                ),
                            ),
                        ),
                    )
                    val heartbeat = launch {
                        while (isActive) {
                            delay(20_000)
                            send(
                                Frame.Text(
                                    json.encodeToString(
                                        Heartbeat(
                                            battery = batteryProvider(),
                                            lanAddress = lanAddressProvider(),
                                        ),
                                    ),
                                ),
                            )
                        }
                    }
                    try {
                        for (frame in incoming) {
                            if (frame is Frame.Text) handle(frame.readText())
                        }
                    } finally {
                        heartbeat.cancel()
                    }
                }
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                // fail soft — fall through to the reconnect delay below
            }
            if (!currentCoroutineContext().isActive) break
            delay(2000)
        }
    }

    private fun handle(text: String) {
        val event = try {
            json.decodeFromString<SignalEvent>(text)
        } catch (e: Exception) {
            return
        }
        when (event.type) {
            "hello-ok" -> event.peers?.forEach {
                setPresence(it.deviceId, it.online ?: true, it.battery, it.publicKey, it.lanAddress)
            }
            "peer-online", "peer-update" -> event.deviceId?.let {
                setPresence(it, true, event.battery, event.publicKey, event.lanAddress)
            }
            // Keep the key and address on the way down: the peer is unreachable now, but
            // the details are still valid when it comes back and save a round trip.
            "peer-offline" -> event.deviceId?.let { setPresence(it, false, null, null, null) }
            "paired" -> _paired.tryEmit(Unit)
        }
    }

    private fun setPresence(
        deviceId: String,
        online: Boolean,
        battery: Int?,
        publicKey: String?,
        lanAddress: String?,
    ) {
        _presence.update { current ->
            val existing = current[deviceId]
            current + (
                deviceId to PeerPresence(
                    online = online,
                    battery = battery ?: existing?.battery,
                    publicKey = publicKey ?: existing?.publicKey,
                    lanAddress = lanAddress ?: existing?.lanAddress,
                )
                )
        }
    }
}
